File: android/src/main/java/com/margelo/nitro/rive/HybridViewModelListProperty.kt
package com.margelo.nitro.rive

import androidx.annotation.Keep
import app.rive.runtime.kotlin.core.ViewModelListProperty
import com.facebook.proguard.annotations.DoNotStrip
import kotlinx.coroutines.flow.map

@Keep
@DoNotStrip
class HybridViewModelListProperty(private val listProperty: ViewModelListProperty) :
  HybridViewModelListPropertySpec(),
  BaseHybridViewModelProperty<Unit> by BaseHybridViewModelPropertyImpl() {
  override val length: Double
    get() = listProperty.size.toDouble()

  private fun requireHybridInstance(instance: HybridViewModelInstanceSpec): HybridViewModelInstance {
    return instance as? HybridViewModelInstance
      ?: throw IllegalArgumentException("Expected HybridViewModelInstance but got ${instance::class.simpleName}")
  }

  override fun getInstanceAt(index: Double): HybridViewModelInstanceSpec? {
    val idx = index.toInt()
    if (idx < 0 || idx >= listProperty.size) return null
    return HybridViewModelInstance(listProperty.elementAt(idx))
  }

  override fun addInstance(instance: HybridViewModelInstanceSpec) {
    val hybridInstance = requireHybridInstance(instance)
    listProperty.add(hybridInstance.viewModelInstance)
  }

  override fun addInstanceAt(instance: HybridViewModelInstanceSpec, index: Double): Boolean {
    val hybridInstance = requireHybridInstance(instance)
    val idx = index.toInt()
    if (idx < 0 || idx > listProperty.size) return false
    listProperty.add(idx, hybridInstance.viewModelInstance)
    return true
  }

  override fun removeInstance(instance: HybridViewModelInstanceSpec) {
    val hybridInstance = requireHybridInstance(instance)
    listProperty.remove(hybridInstance.viewModelInstance)
  }

  override fun removeInstanceAt(index: Double) {
    listProperty.removeAt(index.toInt())
  }

  override fun swap(index1: Double, index2: Double): Boolean {
    return listProperty.swap(index1.toInt(), index2.toInt())
  }

  override fun addListener(onChanged: () -> Unit) {
    listeners.add { _ -> onChanged() }
    ensureValueListenerJob(listProperty.valueFlow.map { })
  }
}

File: ios/HybridViewModelListProperty.swift
import RiveRuntime

class HybridViewModelListProperty: HybridViewModelListPropertySpec, ValuedPropertyProtocol {
  var property: ListPropertyType!
  lazy var helper = PropertyListenerHelper(property: property!)

  init(property: ListPropertyType) {
    self.property = property
    super.init()
  }

  override init() {
    super.init()
  }

  var length: Double {
    Double(property.count)
  }

  func instanceAt(index: Double) throws -> (any HybridViewModelInstanceSpec)? {
    guard let instance = property.instance(at: Int32(index)) else { return nil }
    return HybridViewModelInstance(viewModelInstance: instance)
  }

  private func requireViewModelInstance(_ instance: any HybridViewModelInstanceSpec) throws -> RiveDataBindingViewModel.Instance {
    guard let hybridInstance = instance as? HybridViewModelInstance,
          let viewModelInstance = hybridInstance.viewModelInstance else {
      throw NSError(domain: "HybridViewModelListProperty", code: 1,
                    userInfo: [NSLocalizedDescriptionKey: "Expected HybridViewModelInstance with valid viewModelInstance"])
    }
    return viewModelInstance
  }

  func addInstance(instance: any HybridViewModelInstanceSpec) throws {
    let viewModelInstance = try requireViewModelInstance(instance)
    property.append(viewModelInstance)
  }

  func insertInstance(instance: any HybridViewModelInstanceSpec, index: Double) throws {
    let viewModelInstance = try requireViewModelInstance(instance)
    _ = property.insert(viewModelInstance, at: Int32(index))
  }

  func removeInstance(instance: any HybridViewModelInstanceSpec) throws {
    let viewModelInstance = try requireViewModelInstance(instance)
    property.remove(viewModelInstance)
  }

  func removeInstanceAt(index: Double) throws {
    property.remove(at: Int32(index))
  }

  func swap(index1: Double, index2: Double) throws {
    property.swap(at: UInt32(index1), with: UInt32(index2))
  }

  func addListener(onChanged: @escaping () -> Void) throws {
    try addListener(onChanged: { _ in onChanged() })
  }
}
